File: web/src/services/edgeNode.ts
import { buildQuery } from '../utils/query';
import { request, type AlgorithmPackage } from './api';

// Edge Node types
export interface EdgeNode {
    id: string;
    name: string;
    description?: string;
    endpoint: string;
    max_load: number;
    current_load: number;
    cpu_usage?: number;
    memory_usage?: number;
    media_decode_capacity: number;
    media_encode_capacity: number;
    media_egress_capacity_bps: number;
    media_metrics_ttl_seconds: number;
    status: string; // online, offline, error, disabled
    hal_platform: string;
    engine_version: string;
    hardware_info?: HardwareInfo;
    uptime: number;
    last_heartbeat: string;
    enabled: boolean;
    remark?: string;
    runtime_error?: string;
    created_at: string;
    updated_at: string;

}

// HardwareInfo matches the backend dto.HardwareInfo (total_memory, cpu_cores)
// and also carries legacy display fields used by the frontend pages.
export interface HardwareInfo {
    cpu_model: string;
    gpu_model: string;
    total_memory?: number; // bytes (backend heartbeat field)
    cpu_cores?: number;
    cpu_usage?: number;
    memory_usage?: number;
    // Legacy display fields (mapped from EdgeNode flat fields)
    memory?: string; // display string like "8GB"
    platform?: string;
}

export interface CreateEdgeNodeRequest {
    name: string;
    description?: string;
    endpoint: string;
    max_load: number;
    media_decode_capacity?: number;
    media_encode_capacity?: number;
    media_egress_capacity_bps?: number;
    media_metrics_ttl_seconds?: number;
    remark?: string;
}

export interface UpdateEdgeNodeRequest {
    name?: string;
    description?: string;
    endpoint?: string;
    max_load?: number;
    media_decode_capacity?: number;
    media_encode_capacity?: number;
    media_egress_capacity_bps?: number;
    media_metrics_ttl_seconds?: number;
    enabled?: boolean;
    remark?: string;
}

export interface EdgeNodeListParams {
    page?: number;
    page_size?: number;
    keyword?: string;
    status?: string;
    enabled?: string;
    [key: string]: string | number | undefined;
}

export interface HeartbeatRequest {
    uptime: number;
    current_load: number;
    hardware_info: HardwareInfo;
    installed_algorithms: InstalledAlgorithmInfo[];
    engine_version: string;
}

export interface PendingDeployment {
    algo_package_id: string;
    algo_name: string;
    version: string;
    download_url: string;
    md5: string;
}

export interface InstalledAlgorithmInfo {
    algo_package_id: string;
    algo_name?: string;
    status: string; // installed, failed
    version?: string;
    install_path?: string;
    runtime_status?: string;
    supports_embedding?: boolean;
    supports_face_library?: boolean;
    embedding_capacity?: number;
    error_message?: string;
}

export interface NodeAlgorithm {
    id: string;
    node_id: string;
    algo_package_id: string;
    algo_name: string;
    algo_version: string;
    status: string; // pending, downloading, installed, failed
    download_url: string;
    install_path: string;
    retry_count: number;
    max_retry_count: number;
    last_retry_at: string;
    error_message: string;
    created_at: string;
    updated_at: string;
    algo_package?: AlgorithmPackage;
}

export interface DeployAlgorithmRequest {
    algo_package_id: string;
}

export interface CreateNodeResponse {
    node: EdgeNode;
    token: string;
}

export interface HeartbeatResponse {
    pending_deployments: PendingDeployment[];
}

export interface RecommendNodeResponse {
    recommended_node_id: string;
    node_name: string;
    current_load: number;
    max_load: number;
    load_rate: number;
}

export interface DiskUsageInfo {
    path: string;
    total: number;
    used: number;
    percent: number;
}

export interface EdgeNodeCardSnapshot extends EdgeNode {
    temp?: number;
    active_stream_count?: number;
    decode_slots_used?: number;
    decode_slots_total?: number;
    decode_usage?: number;
    encode_slots_used?: number;
    encode_slots_total?: number;
    encode_usage?: number;
    egress_bps?: number;
    egress_capacity_bps?: number;
    egress_usage?: number;
    accelerator_utilization?: number;
    accelerator_metrics_valid: boolean;
    metrics_received_at?: string;
    disk_usage?: DiskUsageInfo[];
}

export const edgeNodeApi = {
    // CRUD
    list: (params?: EdgeNodeListParams) => {
        const query = buildQuery(params || {});
        return request<{ list: EdgeNodeCardSnapshot[]; total: number; page: number; page_size: number }>(`/edge-nodes${query}`);
    },
    get: (id: string) => request<EdgeNode>(`/edge-nodes/${id}`),
    create: (data: CreateEdgeNodeRequest) =>
        request<CreateNodeResponse>('/edge-nodes', {
            method: 'POST',
            body: JSON.stringify(data),
        }),
    update: (id: string, data: UpdateEdgeNodeRequest) =>
        request<EdgeNode>(`/edge-nodes/${id}`, {
            method: 'PUT',
            body: JSON.stringify(data),
        }),
    delete: (id: string) =>
        request<void>(`/edge-nodes/${id}`, { method: 'DELETE' }),

    // Algorithms
    deployAlgorithm: (nodeId: string, algoPackageId: string) =>
        request<NodeAlgorithm>(`/edge-nodes/${nodeId}/deploy-algo`, {
            method: 'POST',
            body: JSON.stringify({ algo_package_id: algoPackageId }),
        }),
    getNodeAlgorithms: (nodeId: string) =>
        request<NodeAlgorithm[]>(`/edge-nodes/${nodeId}/algorithms`),
    deleteAlgorithm: (nodeId: string, algoPackageId: string) =>
        request<void>(`/edge-nodes/${nodeId}/algorithms/${algoPackageId}`, { method: 'DELETE' }),

    // Heartbeat (used by engine, not admin UI)
    heartbeat: (nodeId: string, token: string, data: HeartbeatRequest) =>
        request<HeartbeatResponse>(`/edge-nodes/${nodeId}/heartbeat`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Bearer ${token}`,
            },
            body: JSON.stringify(data),
        }),

    // Node recommendation (part of AI Vision Tasks)
};

export const recommendNodeApi = {
    recommend: (algoPackageId: string) => {
        const query = buildQuery({ algo_package_id: algoPackageId });
        return request<RecommendNodeResponse>(`/edge-nodes/recommend-node${query}`);
    },
};

File: app/internal/dto/edge_node.go
package dto

import (
	"github.com/niko-admin/niko-admin/internal/pkg/scopes"
)

// EdgeNodeListRequest 边缘节点列表查询参数
type EdgeNodeListRequest struct {
	PageRequest
	Keyword string `form:"keyword"`
	Status  string `form:"status" binding:"omitempty,oneof=online offline error disabled"`
}

// FilterScopes 返回当前请求对应的 GORM 查询范围函数列表
func (r *EdgeNodeListRequest) FilterScopes() []scopes.Scope {
	var sc []scopes.Scope
	if r.Keyword != "" {
		sc = append(sc, scopes.MultiLike([]string{"name", "description", "endpoint", "hal_platform"}, r.Keyword))
	}
	if r.Status != "" {
		sc = append(sc, scopes.Eq("status", r.Status))
	}
	return sc
}

// CreateEdgeNodeRequest 创建边缘节点请求
type CreateEdgeNodeRequest struct {
	Name                   string `json:"name" binding:"required,min=1,max=128"`
	Description            string `json:"description" binding:"max=500"`
	Endpoint               string `json:"endpoint" binding:"required,url,max=255"`
	MaxLoad                int    `json:"max_load" binding:"required,min=1"`
	MediaDecodeCapacity    int    `json:"media_decode_capacity" binding:"min=0"`
	MediaEncodeCapacity    int    `json:"media_encode_capacity" binding:"min=0"`
	MediaEgressCapacityBPS int64  `json:"media_egress_capacity_bps" binding:"min=0"`
	MediaMetricsTTLSeconds int    `json:"media_metrics_ttl_seconds" binding:"min=0"`
	Remark                 string `json:"remark" binding:"max=1000"`
}

// UpdateEdgeNodeRequest 更新边缘节点请求
type UpdateEdgeNodeRequest struct {
	Name                   string `json:"name" binding:"omitempty,min=1,max=128"`
	Description            string `json:"description" binding:"max=500"`
	Endpoint               string `json:"endpoint" binding:"omitempty,url,max=255"`
	MaxLoad                int    `json:"max_load" binding:"omitempty,min=1"`
	MediaDecodeCapacity    *int   `json:"media_decode_capacity" binding:"omitempty,min=0"`
	MediaEncodeCapacity    *int   `json:"media_encode_capacity" binding:"omitempty,min=0"`
	MediaEgressCapacityBPS *int64 `json:"media_egress_capacity_bps" binding:"omitempty,min=0"`
	MediaMetricsTTLSeconds *int   `json:"media_metrics_ttl_seconds" binding:"omitempty,min=0"`

	Enabled *bool  `json:"enabled"`
	Remark  string `json:"remark" binding:"max=1000"`
	Status  string `json:"status" binding:"omitempty,oneof=online offline error disabled"`
}

// HardwareInfo 硬件信息
type HardwareInfo struct {
	CPUModel    string `json:"cpu_model"`
	GPUModel    string `json:"gpu_model"`
	TotalMemory int64  `json:"total_memory" binding:"min=0"`
	CPUCores    int    `json:"cpu_cores" binding:"min=0"`
}

// InstalledAlgorithmInfo 已安装算法信息
type InstalledAlgorithmInfo struct {
	AlgoPackageID       string `json:"algo_package_id" binding:"required,uuid"`
	AlgoName            string `json:"algo_name,omitempty"`
	Version             string `json:"version" binding:"required"`
	InstallPath         string `json:"install_path" binding:"required"`
	Status              string `json:"status" binding:"required"`
	RuntimeStatus       string `json:"runtime_status,omitempty"`
	SupportsEmbedding   bool   `json:"supports_embedding,omitempty"`
	SupportsFaceLibrary bool   `json:"supports_face_library,omitempty"`
	EmbeddingCapacity   int    `json:"embedding_capacity,omitempty" binding:"min=0"`
}

// DiskInfo 磁盘挂载点信息
type DiskInfo struct {
	Path         string  `json:"path"`
	TotalBytes   uint64  `json:"total_bytes"`
	UsedBytes    uint64  `json:"used_bytes"`
	UsagePercent float64 `json:"usage_percent"`
}

// HeartbeatRequest 边缘节点心跳上报请求
type HeartbeatRequest struct {
	Uptime              int64                    `json:"uptime" binding:"min=0"`
	CurrentLoad         int                      `json:"current_load" binding:"min=0"`
	CPUUsage            float64                  `json:"cpu_usage" binding:"min=0,max=100"`
	MemoryUsage         float64                  `json:"memory_usage" binding:"min=0,max=100"`
	EngineVersion       string                   `json:"engine_version" binding:"required"`
	HALPlatform         string                   `json:"hal_platform" binding:"required"`
	HardwareInfo        HardwareInfo             `json:"hardware_info" binding:"required"`
	InstalledAlgorithms []InstalledAlgorithmInfo `json:"installed_algorithms"`
	ActiveStreams       []string                 `json:"active_streams"`
	Status              string                   `json:"status" binding:"omitempty"`
	ErrorMessage        string                   `json:"error_message" binding:"omitempty"`

	// Extended metrics (from C++ Engine MetricsFlattener)
	CPULoad1m    float64    `json:"cpu_load_1m,omitempty"`
	CPULoad5m    float64    `json:"cpu_load_5m,omitempty"`
	CPULoad15m   float64    `json:"cpu_load_15m,omitempty"`
	Disks        []DiskInfo `json:"disks,omitempty"`
	NetRXBytes   int64      `json:"net_rx_bytes,omitempty"`
	NetTXBytes   int64      `json:"net_tx_bytes,omitempty"`
	NetRXSpeed   float64    `json:"net_rx_speed,omitempty"`
	NetTXSpeed   float64    `json:"net_tx_speed,omitempty"`
	ProcessCount int32      `json:"process_count,omitempty"`
	ThreadCount  int32      `json:"thread_count,omitempty"`
	Temperature  float64    `json:"temperature,omitempty"`
}

// PendingDeployment 待下发算法包信息
type PendingDeployment struct {
	AlgoPackageID string `json:"algo_package_id"`
	DownloadURL   string `json:"download_url"`
	MD5           string `json:"md5"`
	ExtractPath   string `json:"extract_path"`
	AlgoName      string `json:"algo_name"`
	Version       string `json:"version"`
}

// HeartbeatResponse 心跳上报响应
type HeartbeatResponse struct {
	PendingDeployments []PendingDeployment `json:"pending_deployments"`
}

// DeployAlgorithmRequest 部署算法请求
type DeployAlgorithmRequest struct {
	AlgoPackageID string `json:"algo_package_id" binding:"required,uuid"`
}

// DeployAlgorithmResponse 部署算法响应
type DeployAlgorithmResponse struct {
	DeploymentID string `json:"deployment_id"`
	Message      string `json:"message"`
}
